### PR TITLE
Introduce a LayoutPath type

### PR DESCRIPTION
**Pull request by glyn**

### CODE REVIEW

File: pkg/v1/remote/write_test.go [hunk 1]
```diff
@@ -1067,6 +1067,6 @@ func TestWriteIndex(t *testing.T) {
 	}
 
 	if err := WriteIndex(tag, idx, authn.Anonymous, http.DefaultTransport); err != nil {
-		t.Errorf("WriteIndex() = %v", err)
+		t.Errorf("writeIndex() = %v", err)
```
Comment (jonjohnsonjr) [inline pkg/v1/remote/write_test.go]: Looks like this one slipped in accidentally.

File: pkg/v1/layout/write_test.go [hunk 2]
```diff
@@ -183,7 +203,11 @@ func TestOptions(t *testing.T) {
 		WithURLs(urls),
 		WithPlatform(platform),
 	}
-	idx, err := AppendImage(tmp, img, options...)
+	err = temp.AppendImage(img, options...)
```
Comment (jonjohnsonjr) [inline pkg/v1/layout/write_test.go]: nit: `if err := tempt.AppendImage(img, options...); err != nil {`

File: pkg/v1/layout/read.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+// limitations under the License.
+
+package layout
+
+import (
+	"io/ioutil"
+	"path/filepath"
+)
+
+// Read reads an OCI image layout at path and constructs a LayoutPath.
```
Comment (jonjohnsonjr) [inline pkg/v1/layout/read.go:22]: It seems like almost every time I do a `Read` I'll immediatelly follow that with a call to `ImageIndex()`. I wonder if there's a clean way to combine the two into one call to make this easier to use.

File: pkg/v1/layout/read.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+func Read(path string) (LayoutPath, error) {
+	// TODO: check oci-layout exists
+
+	_, err := ioutil.ReadFile(filepath.Join(path, "index.json"))
```
Comment (jonjohnsonjr) [inline pkg/v1/layout/read.go]: Seems wasteful to read this without doing anything with it. You could just stat the file or stash it away for a later call to `ImageIndex`?

I'd also want some way to open a LayoutPath for writing without having to write an empty index first. Maybe we could change this to `Open` (or add another method) that will initialize to an empty.Index if not found.

File: pkg/v1/remote/write_test.go (hunk 1 above)
Comment (glyn) [reply pkg/v1/remote/write_test.go]: Whoops! Thanks for spotting. Fixed.

File: pkg/v1/layout/write_test.go (hunk 2 above)
Comment (glyn) [reply pkg/v1/layout/write_test.go]: Changed as suggested.

File: pkg/v1/layout/read.go (hunk 3 above)
Comment (glyn) [reply pkg/v1/layout/read.go:22]: I added a convenience function `ImageIndex`. Hope you don't mind it sharing a name with the corresponding method.

File: pkg/v1/layout/read.go (hunk 4 above)
Comment (glyn) [reply pkg/v1/layout/read.go]: > Seems wasteful to read this without doing anything with it. You could just stat the file or stash it away for a later call to `ImageIndex`?

Changed to use stat.

> 
> I'd also want some way to open a LayoutPath for writing without having to write an empty index first. Maybe we could change this to `Open` (or add another method) that will initialize to an empty.Index if not found.

Ran out of time today. Need to think about this on Monday, but please feel free to merge this and change as necessary before then.

**Review by jonjohnsonjr — APPROVED** [review]
LGTM, I think we might want to expose `WriteBlob`, `WriteImage`, and `WriteIndex` eventually, but I don't have a use for it now so it's better to be conservative with our API surface. Thanks!

**Pull request merged**